codalab/lib/upload_manager.py

```diff
@@ -258,3 +261,210 @@ def cleanup_existing_contents(self, bundle):
+        """
+        return (
+            parse_linked_bundle_url(path)
+            .bundle_path_sas_url(permission='rw', **kwargs)
+            .split('?')[-1]
+        )
+
+    def get_index_sas_token(self, path, **kwargs):
+        """
+        Get SAS token of the index file with read and write permission. Used for uploading.
+        """
+        return (
+            parse_linked_bundle_url(path)
+            .index_path_sas_url(permission='rw', **kwargs)
+            .split('?')[-1]
+        )
+
+    def get_bundle_signed_url(self, path, **kwargs):
+        """
+        Get signed url for the bundle path
+        """
+        return parse_linked_bundle_url(path).bundle_path_signed_url(**kwargs)
+
+    def get_bundle_index_url(self, path, **kwargs):
+        return parse_linked_bundle_url(path).index_path_signed_url(**kwargs)
+
+
+class ClientUploadManager(object):
+    """
+    Upload Manager for CLI client. Handle file upload for CLI client.
+    """
+
+    def __init__(self, json_api_client, stdout, stderr):
+        self._client = json_api_client
+        self.stdout = stdout
+        self.stderr = stderr
+
+    def upload_to_bundle_store(
+        self,
+        bundle: Dict,
+        packed_source: Dict,
+        use_azure_blob_beta: bool,
+        destination_bundle_store=None,
+    ):
+
+        # By pass server upload:
+        # 1. If the user specify `-a`, upload to Azure blob storage
+        # 2. If the user specify `--store` and blob storage is on Azure or GCS
+        bypass_server = True if use_azure_blob_beta else False
+        bundle_store_uuid = None
+        if destination_bundle_store is not None and destination_bundle_store != '':
+            storage_info = self._client.fetch_one(
+                'bundle_stores',
+                params={
+                    'name': destination_bundle_store,
+                    'include': ['uuid', 'storage_type', 'url'],
+                },
+            )
+            bundle_store_uuid = storage_info['uuid']
+            if storage_info['storage_type'] in (
+                StorageType.AZURE_BLOB_STORAGE.value,
+                StorageType.GCS_STORAGE.value,
+            ):
+                bypass_server = True
+
+        if bypass_server:
+            # Mimic the rest server behavior
+            # decided the bundle type (file/directory) and decide whether need to unpack
+            source_ext = zip_util.get_archive_ext(packed_source['filename'])
+            if packed_source['should_unpack'] and zip_util.path_is_archive(
+                packed_source['filename']
+            ):
+                unpack_before_upload = True
+                is_dir = source_ext in zip_util.ARCHIVE_EXTS_DIR
+            else:
+                unpack_before_upload = False
+                is_dir = False
+
+            params = {'need_sas': True, 'is_dir': is_dir}
+            data = self._client.update_bundle_locations(bundle['id'], bundle_store_uuid, params)[
```

Comment:
Even if bypass_server is False, we should call update_bundle_locations nevertheless. If bundle_conn_str / index_conn_str are None, then we should just call the regular PUT endpoint to upload the bundle through the server.

Reply:
> Even if bypass_server is False, we should call update_bundle_locations nevertheless. If bundle_conn_str / index_conn_str are None, then we should just call the regular PUT endpoint to upload the bundle through the server.

Fixed. Now all the uploads will first try to do bypass server upload. (except the user specify `--store {some_disk_storage}`